File: python/techai_webutils/src/techai_webutils/core/interfaces/metrics.py
"""Metrics interface for application-level metric recording.

Defines the abstract contract for metrics implementations.
Foundation implementations satisfy this interface; consumers depend only on it.
"""

from __future__ import annotations

from abc import ABC, abstractmethod


class MetricsProvider(ABC):
    """Abstract metrics provider for counters, histograms, and gauges.

    Implementations: Prometheus (default).
    All service code depends on this interface, never on a concrete metrics library.
    """

    @abstractmethod
    def counter(self, name: str, help_text: str, labels: list[str] | None = None) -> MetricCounter:
        """Create or retrieve a counter metric."""
        ...

    @abstractmethod
    def histogram(
        self,
        name: str,
        help_text: str,
        labels: list[str] | None = None,
        buckets: list[float] | None = None,
    ) -> MetricHistogram:
        """Create or retrieve a histogram metric."""
        ...

    @abstractmethod
    def gauge(self, name: str, help_text: str, labels: list[str] | None = None) -> MetricGauge:
        """Create or retrieve a gauge metric."""
        ...


class MetricCounter(ABC):
    """Abstract counter metric."""

    @abstractmethod
    def inc(self, value: float = 1.0, **labels: str) -> None:
        """Increment the counter."""
        ...

    def increment(self, value: float = 1.0, **labels: str) -> None:
        """Alias for :meth:`inc` for readability."""
        self.inc(value, **labels)


class MetricHistogram(ABC):
    """Abstract histogram metric."""

    @abstractmethod
    def observe(self, value: float, **labels: str) -> None:
        """Record an observation."""
        ...


class MetricGauge(ABC):
    """Abstract gauge metric."""

    @abstractmethod
    def set(self, value: float, **labels: str) -> None:
        """Set the gauge to a value."""
        ...

    @abstractmethod
    def inc(self, value: float = 1.0, **labels: str) -> None:
        """Increment the gauge."""
        ...

    @abstractmethod
    def dec(self, value: float = 1.0, **labels: str) -> None:
        """Decrement the gauge."""
        ...

File: go/core/interfaces/metrics.go
package interfaces

import "net/http"

// Metrics provides application-level metric recording.
//
// Implementations: Prometheus (default) or any metrics backend that supports
// counters, histograms, and gauges.
//
// All foundation and service code depends on this interface, never on a
// concrete metrics library. Swap implementations via the metrics.New factory.
type Metrics interface {
	// Counter returns a counter that can be incremented.
	Counter(name, help string, labels ...string) Counter

	// Histogram returns a histogram for recording distributions.
	Histogram(name, help string, buckets []float64, labels ...string) Histogram

	// Gauge returns a gauge for recording point-in-time values.
	Gauge(name, help string, labels ...string) Gauge

	// Handler returns an HTTP handler that serves the metrics endpoint.
	Handler() http.Handler
}

// Counter is an incrementing metric.
type Counter interface {
	// Inc increments the counter by 1 with the given label values.
	Inc(labelValues ...string)

	// Add increments the counter by the given value.
	Add(value float64, labelValues ...string)
}

// Histogram records value distributions.
type Histogram interface {
	// Observe records a value with the given label values.
	Observe(value float64, labelValues ...string)
}

// Gauge records point-in-time values.
type Gauge interface {
	// Set sets the gauge to the given value.
	Set(value float64, labelValues ...string)

	// Inc increments the gauge by 1.
	Inc(labelValues ...string)

	// Dec decrements the gauge by 1.
	Dec(labelValues ...string)
}
